art/attacks/hop_skip_jump.py

```diff
@@ -0,0 +1,513 @@
+        :param y_p: The predicted label of x.
+        :type y_p: `int`
+        :param init_pred: The predicted label of the initial image.
+        :type init_pred: `int`
+        :param adv_init: Initial array to act as an initial adversarial example.
+        :type adv_init: `np.ndarray`
+        :param clip_min: Minimum value of an example.
+        :type clip_min: `float`
+        :param clip_max: Maximum value of an example.
+        :type clip_max: `float`
+        :return: an adversarial example.
+        :rtype: `np.ndarray`
+        """
+        # First, create an initial adversarial sample
+        initial_sample = self._init_sample(x, y, y_p, init_pred, adv_init, clip_min, clip_max)
+
+        # If an initial adversarial example is not found, then return the original image
+        if initial_sample is None:
+            return x
+
+        # If an initial adversarial example found, then go with hopskipjump attack
+        x_adv = self._attack(initial_sample[0], x, initial_sample[1], clip_min, clip_max)
+
+        return x_adv
+
+    def _init_sample(self, x, y, y_p, init_pred, adv_init, clip_min, clip_max):
+        """
+        Find initial adversarial example for the attack.
+
+        :param x: An array with 1 original input to be attacked.
+        :type x: `np.ndarray`
+        :param y: If `self.targeted` is true, then `y` represents the target label.
+        :type y: `int`
+        :param y_p: The predicted label of x.
+        :type y_p: `int`
+        :param init_pred: The predicted label of the initial image.
+        :type init_pred: `int`
+        :param adv_init: Initial array to act as an initial adversarial example.
+        :type adv_init: `np.ndarray`
+        :param clip_min: Minimum value of an example.
+        :type clip_min: `float`
+        :param clip_max: Maximum value of an example.
+        :type clip_max: `float`
+        :return: an adversarial example.
+        :rtype: `np.ndarray`
+        """
+        nprd = np.random.RandomState()
+        initial_sample = None
+
+        if self.targeted:
+            # Attack satisfied
+            if y == y_p:
+                return None
+
+            # Attack unsatisfied yet and the initial image satisfied
+            if adv_init is not None and init_pred == y:
+                return adv_init.astype(NUMPY_DTYPE), init_pred
+
+            # Attack unsatisfied yet and the initial image unsatisfied
+            for _ in range(self.init_size):
+                random_img = nprd.uniform(clip_min, clip_max, size=x.shape).astype(x.dtype)
+                random_class = np.argmax(self.classifier.predict(np.array([random_img])), axis=1)[0]
+
+                if random_class == y:
+                    # Binary search to reduce the l2 distance to the original image
+                    tmp_norm = self.norm
+                    self.norm = 2
+                    random_img = self._binary_search(current_sample=random_img, original_sample=x, target=y,
```

Comment:
It looks like `self.norm` is set to `2` and the actual norm stored in `tmp_nrom` because in this line `_binary_search` should run with `norm=2` for `self.norm=2` and `self.norm=np.inf` cases. Would it be possible to define the `norm` that `_binary_search` should use as an argument of `_binary_search` instead of using `self.norm`?

Reply:
Solved.